src/java_bytecode/java_string_literals.cpp

```diff
@@ -0,0 +1,211 @@
+    const address_of_exprt array_pointer(
+      index_exprt(array_expr, from_integer(0, java_int_type())));
+
+    const std::size_t data_nb = jls_struct.component_number("data");
+    literal_init.operands()[data_nb] = array_pointer;
+
+    // Associate array with pointer
+    symbolt return_symbol;
+    return_symbol.name = escaped_symbol_name_with_prefix + "_return_value";
+    return_symbol.base_name = escaped_symbol_name + "_return_value";
+    return_symbol.pretty_name =
+      escaped_symbol_name.length() > 10
+        ? escaped_symbol_name.substr(0, 10) + "..._return_value"
+        : escaped_symbol_name + "_return_value";
+    return_symbol.mode = ID_java;
+    return_symbol.is_type = false;
+    return_symbol.is_lvalue = true;
+    return_symbol.is_static_lifetime = true;
+    return_symbol.is_state_var = true;
+    return_symbol.value = make_function_application(
+      ID_cprover_associate_array_to_pointer_func,
+      {array_symbol.value, array_pointer},
+      java_int_type(),
+      symbol_table);
+    return_symbol.type = return_symbol.value.type();
+    if(symbol_table.add(return_symbol))
+      throw "failed to add return symbol to symbol table";
+
+    // Initialise the string literal with
+    // (&..._return_value, { actual initializer} ) to ensure that used-global
+    // variable analysis sees that if the literal is required, so is the
+    // return_symbol that executes the pointer <-> array association.
+    exprt comma_expr(ID_comma, literal_init.type());
+    comma_expr.copy_to_operands(return_symbol.symbol_expr());
+    comma_expr.copy_to_operands(literal_init);
+    new_symbol.value = comma_expr;
+  }
+  else if(jls_struct.components().size()>=1 &&
+          jls_struct.components()[0].get_name()=="@java.lang.Object")
+  {
+    // Case where something defined java.lang.String, so it has
+    // a proper base class (always java.lang.Object in practical
+    // JDKs seen so far)
+    struct_exprt literal_init(new_symbol.type);
+    literal_init.move_to_operands(jlo_init);
+    for(const auto &comp : jls_struct.components())
+    {
+      if(comp.get_name()=="@java.lang.Object")
+        continue;
+      // Other members of JDK's java.lang.String we don't understand
+      // without string-refinement. Just zero-init them; consider using
+      // test-gen-like nondet object trees instead.
+      literal_init.copy_to_operands(
+        zero_initializer(comp.type(), string_expr.source_location(), ns));
+    }
+    new_symbol.value = literal_init;
+  }
+  else if(jls_struct.get_bool(ID_incomplete_class))
+  {
+    // Case where java.lang.String was stubbed, and so directly defines
+    // @class_identifier and @lock:
+    new_symbol.value = jlo_init;
+  }
+
+  bool add_failed = symbol_table.add(new_symbol);
+  INVARIANT(
+    !add_failed,
+    "string literal symbol was already checked not to be "
+    "in the symbol table, so adding it should succeed");
+
+  return new_symbol.symbol_expr();
+}
```

Comment:
This function is way too long (but I suppose it would be acceptable to put it here as a move-only commit).